packages/labs/src/components/timezone-select/timezoneSelect.tsx

```diff
@@ -0,0 +1,509 @@
+
+    /**
+     * Format to use when displaying the selected (or default) timezone within the target element.
+     * @default TimezoneDisplayFormat.OFFSET
+     */
+    targetDisplayFormat?: TimezoneDisplayFormat;
+
+    /**
+     * A space-delimited list of class names to pass along to the target element.
+     * This prop is ignored when a `targetRenderer` is provided.
+     */
+    targetClassName?: string;
+
+    /**
+     * Whether this component is non-interactive.
+     * @default false
+     */
+    disabled?: boolean;
+
+    /**
+     * Text to show when no timezone has been selected and there is no default.
+     * @default GMT timezone formatted according to `targetDisplayFormat`
+     */
+    placeholder?: string;
+
+    /**
+     * Props to spread to `Popover`. Note that `content` cannot be changed.
+     */
+    popoverProps?: Partial<IPopoverProps> & object;
+}
+
+export type ITimezoneSelectTargetRenderer = (targetProps: ITimezoneSelectTargetProps) => JSX.Element | null;
+
+export interface ITimezoneSelectTargetProps {
+    /** The currently selected timezone. */
+    value: string | undefined;
+
+    /** Initial timezone, when none is selected. */
+    defaultValue: string | undefined;
+
+    /** Display version of the currently selected timezone. */
+    displayValue: string | undefined;
+
+    /** Display version of the default timezone. */
+    defaultDisplayValue: string | undefined;
+
+    /** Placeholder for when no timezone has been selected and there is no default. */
+    placeholder: string;
+
+    /** Whether the target is intended to be non-interactive. */
+    disabled: boolean;
+
+    /** Date to use when determining timezone offsets. */
+    date: Date;
+}
+
+export type TimezoneDisplayFormat = "offset" | "abbreviation" | "name";
+export const TimezoneDisplayFormat = {
+    ABBREVIATION: "abbreviation" as "abbreviation",
+    NAME: "name" as "name",
+    OFFSET: "offset" as "offset",
+};
+
+export interface ITimezoneSelectState {
+    date?: Date;
+    value?: string;
+    query?: string;
+}
+
+interface ITimezoneItem {
+    key: string;
+    text: string;
+    label: string;
+    iconName: IconName | undefined;
+    timezone: string;
+}
+
+const TypedSelect = Select.ofType<ITimezoneItem>();
+
+const PLACEHOLDER_TIMEZONE = "GMT";
+
+@PureRender
+export class TimezoneSelect extends AbstractComponent<ITimezoneSelectProps, ITimezoneSelectState> {
+    public static displayName = "Blueprint.TimezoneSelect";
+
+    public static defaultProps: Partial<ITimezoneSelectProps> = {
+        defaultToLocalTimezone: false,
+        disabled: false,
+        popoverProps: {},
+        showLocalTimezone: true,
+    };
+
+    private timezones: ITimezoneItem[];
+    private timezoneToQueryCandidates: { [timezone: string]: string[] };
+    private initialTimezones: ITimezoneItem[];
+
+    constructor(props: ITimezoneSelectProps, context?: any) {
+        super(props, context);
+
+        const date = props.date || new Date();
```

Comment:
deconstruct with default value. more type-safe than `||` pattern.
```
const { date = new Date() } = props;
```

Reply:
How do you feel about the `||` pattern in react components?
```tsx
           <Button
                className={classes}
                text={displayValue || defaultDisplayValue || placeholder}
                rightIconName="caret-down"
                disabled={disabled}
            />
```

Reply:
i'm intrigued by the conciseness but it's honestly quite unsettling: classic abuse of a js language "feature," which we've tended to avoid in this codebase.